bump tokenizers, fix added tokens fast

ArthurZucker opened the pull request:
# What does this PR do?
Since tokenizer is being released, we can now fix the way tokens are added

ArthurZucker committed:
update based on tokenizers release

ArthurZucker committed:
update

ArthurZucker committed:
nits

ArthurZucker committed:
update

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32535). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker committed:
revert re addition

ArthurZucker committed:
don't break that yet

ArthurZucker committed:
fmt

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into fix-added-tokens-fast

ArthurZucker committed:
revert unwanted

ArthurZucker committed:
update tokenizers version

ArthurZucker committed:
update dep table

ArthurZucker committed:
update

ArthurZucker committed:
update in conversion script as well

ArthurZucker committed:
some fix

ArthurZucker committed:
revert

ArthurZucker committed:
fully revert

ArthurZucker committed:
fix training

ArthurZucker committed:
remove set trace

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into fix-added-tokens-fast

ArthurZucker committed:
fixup

ArthurZucker committed:
update

LysandreJik commented on src/transformers/convert_slow_tokenizer.py:
```diff
@@ -602,33 +602,12 @@ def tokenizer(self, proto):
-                    tokenizer.add_special_tokens(tokens)
-                else:
-                    tokenizer.add_tokens(tokens)
+        tokenizer.add_tokens(
+            [
+                AddedToken(token, normalized=False, special=special)
+                for id, token, special in sorted(spm_added_tokens, key=lambda x: x[0])
+            ]
+        )
```
Much cleaner!

LysandreJik reviewed (approved):
Nice, the diff is nice to see!

ArthurZucker commented:
Yeah still a few bugs to fix tho :( getting this ready for next release!

ArthurZucker committed:
t push
Merge branch 'main' of github.com:huggingface/transformers into fix-added-tokens-fast

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into fix-added-tokens-fast
t push

itazap commented on src/transformers/tokenization_utils_fast.py:
```diff
@@ -191,27 +184,7 @@ def __init__(self, *args, **kwargs):
             token for token in self.all_special_tokens_extended if token not in encoder and token not in tokens_to_add
```
```suggestion
tokens_to_add += [
          AddedToken(token, special=True) if isinstance(token, str) else token
          for token in self.all_special_tokens_extended
          if token not in encoder and token not in tokens_to_add
      ]```

itazap commented on src/transformers/tokenization_utils_fast.py:
```diff
@@ -191,27 +184,7 @@ def __init__(self, *args, **kwargs):
             token for token in self.all_special_tokens_extended if token not in encoder and token not in tokens_to_add
         ]
 
```
ensures tokens are added as special if they are in all_special_tokens_extended (fixes test failure with `additional_special_tokens` param passed

itazap replied:
also somehow `all_special_tokens_extended` (using `special_tokens_map_extended`) has the `<mask>` token as `special=False` which causes a test failure

ArthurZucker replied:
Mmmm but I'd rather we just take them from `added_tokens_decoder` raw, as it only stores added tokens!

itazap commented on src/transformers/tokenization_utils_fast.py:
```diff
@@ -191,27 +184,7 @@ def __init__(self, *args, **kwargs):
-            # individual tokens would repeatedly rebuild a trie, which can be slow.
-            is_last_special = None
-            tokens = []
-            special_tokens = self.all_special_tokens
```
this used to ensure that tokens get added as special if they are in `all_special_tokens`. It even overwrote the default `special=False` param when creating an `AddedToken` such as the mask token in fast models (like here: https://github.com/huggingface/transformers/blob/2b789f27f383435b8db2fee3d10b0a1358c0c234/src/transformers/models/layoutxlm/tokenization_layoutxlm_fast.py#L233-L234 

But I'm not sure why only the mask token gets added like this in so many models and if we should just pass `special=True` there in each model?

ArthurZucker replied:
yeah we should make them special

ArthurZucker replied:
Actually you are right, I kep the logic to check special / change special if in all special tokens

ArthurZucker replied:
All tests pass thanks to you

ArthurZucker committed:
Merge branch 'main' of github.com:huggingface/transformers into fix-added-tokens-fast

ArthurZucker committed:
update
